fix(compiler): pipes using DI not working in blocks

crisbeto opened the pull request:
Fixes that the new block syntax was generating instructions in the wrong order which meant that pipes were being declared too early. This meant that if the block is first in the template, any pipes used in it won't be able to inject things like `ChangeDetectorRef`.

These changes update the compiler and add a bunch of tests to ensure that pipes work as expected.

Fixes #52102.

pkozlowski-opensource commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_deferred/deferred_when_with_pipe_template.js:
```diff
@@ -3,11 +3,11 @@ function MyApp_Template(rf, ctx) {
     $r3$.ɵɵtemplate(1, MyApp_Defer_1_Template, 1, 0);
     $r3$.ɵɵdefer(2, 1);
-    $r3$.ɵɵpipe(3, "testPipe");
+    $r3$.ɵɵpipe(4, "testPipe");
```
This change looks suspicious for me - why do we change the index of a pipe here?

crisbeto replied:
Capturing the DM discussion: we pre-allocate one slot for the deferred block runtime state.

crisbeto commented:
Caretaker note: this one passed a presubmit earlier. I had to push again to resolve merge conflicts in a test.

crisbeto committed:
fix(compiler): pipes using DI not working in blocks

Fixes that the new block syntax was generating instructions in the wrong order which meant that pipes were being declared too early. This meant that if the block is first in the template, any pipes used in it won't be able to inject things like `ChangeDetectorRef`.

These changes update the compiler and add a bunch of tests to ensure that pipes work as expected.

Fixes #52102.
